clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -488,18 +489,72 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
   return Store;
 }
 
+static CallInst *CreateBuiltinCallWithAttr(CodeGenFunction &CGF, StringRef Name,
+                                           llvm::Function *FPBuiltinF,
+                                           ArrayRef<Value *> Args,
+                                           unsigned ID) {
+  llvm::CallInst *CI = CGF.Builder.CreateCall(FPBuiltinF, Args);
+  llvm::AttributeList AttrList;
+  if (Name == "sincos")
+    CGF.CGM.getFPAccuracyFuncAttributes(Name, AttrList, ID, Args[0]->getType());
+  else
+    CGF.CGM.getFPAccuracyFuncAttributes(Name, AttrList, ID,
+                                        FPBuiltinF->getReturnType());
+  CI->setAttributes(AttrList);
+  return CI;
+}
+
+Function *getIntrinsic(CodeGenFunction &CGF, llvm::Value *Src0,
+                       unsigned FPIntrinsicID, unsigned IntrinsicID,
+                       bool HasAccuracyRequirement) {
+  return HasAccuracyRequirement
+             ? CGF.CGM.getIntrinsic(FPIntrinsicID, Src0->getType())
+             : CGF.CGM.getIntrinsic(IntrinsicID, Src0->getType());
+}
+
+bool hasAccuracyRequirement(CodeGenFunction &CGF, StringRef Name) {
+  bool HasRequirement = false;
+  if (!CGF.getLangOpts().FPAccuracyVal.empty())
+    HasRequirement = true;
+  if (!CGF.getLangOpts().FPAccuracyFuncMap.empty()) {
+    auto FuncMapIt = CGF.getLangOpts().FPAccuracyFuncMap.find(Name.str());
+    if (FuncMapIt != CGF.getLangOpts().FPAccuracyFuncMap.end())
+      HasRequirement = true;
+  }
+  return HasRequirement;
+}
+
 // Emit a simple mangled intrinsic that has 1 argument and a return type
 // matching the argument type. Depending on mode, this may be a constrained
-// floating-point intrinsic.
-static Value *emitUnaryMaybeConstrainedFPBuiltin(CodeGenFunction &CGF,
-                                const CallExpr *E, unsigned IntrinsicID,
-                                unsigned ConstrainedIntrinsicID) {
+// or an fpbuiltin floating-point intrinsic.
+static Value *emitUnaryMaybeConstrainedFPBuiltin(
+    CodeGenFunction &CGF, const CallExpr *E, unsigned IntrinsicID,
+    unsigned ConstrainedIntrinsicID,
+    unsigned FPAccuracyIntrinsicID = Intrinsic::not_intrinsic) {
   llvm::Value *Src0 = CGF.EmitScalarExpr(E->getArg(0));
-
+  if (FPAccuracyIntrinsicID != Intrinsic::not_intrinsic) {
+    if (CGF.CGM.getCodeGenOpts().FPAccuracy) {
+      if (CGF.getLangOpts().MathErrno) {
+        DiagnosticsEngine &Diags = CGF.CGM.getDiags();
+        Diags.Report(E->getBeginLoc(),
+                     diag::err_incompatible_fp_accuracy_options);
+      } else {
+        unsigned BuiltinID = CGF.getCurrentBuiltinID();
```

Comment:
I feel like this else has some unnecessary temporary variables used just once.  We can just call the function directly instead of using a variable here.